src/server/pkg/obj/amazon_client.go

```diff
@@ -37,16 +44,143 @@ type amazonClient struct {
 	uploader               *s3manager.Uploader
 }
 
-func newAmazonClient(bucket string, cloudfrontDistribution string, id string, secret string, token string, region string) (*amazonClient, error) {
-	config := &aws.Config{
+type vaultCredentialsProvider struct {
+	vaultClient *vault.Client // client used to retrieve S3 creds from vault
+	vaultRole   string        // get vault creds from: /aws/creds/<vaultRole>
+
+	// ID, duration, and last renew time of the vault lease that governs the most
+	// recent AWS secret issued to this pachd instance (and a mutex protecting
+	// them)
+	leaseMu        sync.Mutex
+	leaseID        string
+	leaseLastRenew time.Time
+	leaseDuration  time.Duration
+}
+
+// updateLease extracts the duration of the lease governing 'secret' (an AWS
+// secret). IIUC, because the AWS backend issues dynamic secrets, there is no
+// tokens associated with them, and vaultSecret.TokenTTL can be ignored
+func (v *vaultCredentialsProvider) updateLease(secret *vault.Secret) {
+	v.leaseMu.Lock()
+	defer v.leaseMu.Unlock()
+	v.leaseID = secret.LeaseID
+	v.leaseLastRenew = time.Now()
+	v.leaseDuration = time.Duration(secret.LeaseDuration) * time.Second
+}
+
+// Retrieve returns nil if it successfully retrieved the value.  Error is
+// returned if the value were not obtainable, or empty.
+func (v *vaultCredentialsProvider) Retrieve() (credentials.Value, error) {
+	var emptyCreds, result credentials.Value // result
+
+	// retrieve AWS creds from vault
+	vaultSecret, err := v.vaultClient.Logical().Read(path.Join("aws", "creds", v.vaultRole))
+	if err != nil {
+		return emptyCreds, fmt.Errorf("could not retrieve creds from vault: %v", err)
+	}
+	accessKeyIface, accessKeyOk := vaultSecret.Data["access_key"]
+	awsSecretIface, awsSecretOk := vaultSecret.Data["secret_key"]
+	if !accessKeyOk || !awsSecretOk {
+		return emptyCreds, fmt.Errorf("aws creds not present in vault response")
+	}
+
+	// Convert access key & secret in response to strings
+	result.AccessKeyID, accessKeyOk = accessKeyIface.(string)
+	result.SecretAccessKey, awsSecretOk = awsSecretIface.(string)
+	if !accessKeyOk || !awsSecretOk {
+		return emptyCreds, fmt.Errorf("aws creds in vault response were not both strings (%T and %T)", accessKeyIface, awsSecretIface)
+	}
+
+	// update the lease values in 'v', and spawn a goroutine to renew the lease
+	v.updateLease(vaultSecret)
+	go func() {
+		for {
+			// renew at half the lease duration or one day, whichever is greater
+			// (lease must expire eventually)
+			renewInterval := v.leaseDuration
```

Comment:
Should this be holding the mutex will it accesses v.leaseDuration? It looks like that could possibly race with the code above. But I'm not quite sure.

Reply:
Good catch. I'll rewrite this code a bit